server/lib/db/pg-error-parser.ts

```diff
@@ -0,0 +1,42 @@
+import { SYNTAX_ERROR_OR_ACCESS_RULE_VIOLATION } from './pg-error-codes'
+
+export default function parseError(query: string, error: any): Error {
+  if (
+    error.code &&
+    error.code.substring(0, 2) === SYNTAX_ERROR_OR_ACCESS_RULE_VIOLATION.substring(0, 2) &&
+    error.position
+  ) {
+    return parseSyntaxError(query, error)
+  }
+  return new Error(`${error.message} in query ${query}`)
+}
+
+function parseSyntaxError(queryText: string, error: any): Error {
+  let queryLines = queryText.split('\n')
+
+  let cursor = 0
+  let lineNumber = 0
+  let relativePosition = 0
+  let foundPosition = false
+  for (let i = 0; i < queryLines.length && !foundPosition; i++) {
+    lineNumber += 1
+    // We need to count the new line character
+    cursor += 1
```

Comment:
This seems like it'd be an off-by-one thing because the lines do not start with a newline, they end with it.

This whole logic seems a bit convoluted for what you're actually trying to do here, as well. We know the position is within the range `[0, length]`, we want to know what line it's on. Given an array of lines (which you got above from the `split`), you can iterate through the lines until the summed length is `> error.position`. Once that is the case, the position is on the previous line.

```ts
const position = Number(error.position)
if (isNaN(position)) {
  return new Error(`${error.message} in query: \n${queryLines.join('\n')}`)
}

let curLine = 0
let summedLength = 0
do {
  summedLength += queryLines[curLine].length
  curLine += 1
} while (curLine < queryLines.length && summedLength < position)

if (summedLength >= position) {
  const foundLine = curLine - 1
  const relativePosition = position - (summedLength - queryLines[foundLine].length)
  const caretLine = '^'.padStart(relativePosition + 1)
  queryLines.splice(curLine, 0, caretLine)
}

return new Error(`${error.message} in query: \n${queryLines.join('\n')}`)
```